src/Data/HashMap/Linear.hs

```diff
@@ -280,6 +432,136 @@ instance Consumable (HashMap k v) where
   consume :: HashMap k v #-> ()
   consume (HashMap _ arr) = consume arr
 
+instance Data.Functor (HashMap k) where
+  fmap f (HashMap c arr) =
+    HashMap c $
+      Data.fmap
+        (\case
+          Nothing -> Nothing
+          Just (RobinVal p k v) -> Just (RobinVal p k (f v))
+        )
+        arr
+
+instance Prelude.Semigroup (HashMap k v) where
+  (<>) = error "Prelude.<>: invariant violation, unrestricted HashMap"
+
+instance Keyed k => Semigroup (HashMap k v) where
+  (<>) = union
+
+-- # Internal library
+--------------------------------------------------
+
 _debugShow :: (Show k, Show v) => HashMap k v #-> String
 _debugShow (HashMap _ robinArr) =
   Array.toList robinArr & \(Ur xs) -> show xs
+
+-- | When using Robin-Hood hashing, the beginning of the array
+-- can contain elements "spilled over" from the end of the array.
+--
+-- This function finds the number of spillovers. The spillovers can
+-- be determined by the length of the prefix of the array where
+-- PSL > index.
+numSpillovers :: RobinArr k v #-> (RobinArr k v, Ur Int)
+numSpillovers arr =
```

Comment:
Technically this implementation is correct because spillovers form a contiguous segment (which is easy to prove). Just comment that.